File: Dynamic_Programming/Python/Trapping_Water.py
# Trapping Rain Water 
# Author: Pavan Kumar Paluri
# Leetcode Question: https://leetcode.com/problems/trapping-rain-water/

class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) ==0:
            # No towers
            return 0
        ans = 0
        left_arr = [None]*len(height)
        right_arr = [None]*len(height)
        left_arr[0] = height[0]
        for i in range(1, len(height)):
            left_arr[i] = max(height[i], left_arr[i-1])
        right_arr[len(height)-1] = height[len(height)-1]
        for j in range(len(height)-2, -1, -1):
            right_arr[j] = max(height[j], right_arr[j+1])
        for k in range(0, len(height), 1):
            ans += min(left_arr[k], right_arr[k])-height[k]
        return ans 

File: Dynamic_Programming/C++/trapping_water.cpp
// Trapping Rain water - Leetcode Hard 
// Author: Pavan Kumar Paluri
// Leetcode Question: https://leetcode.com/problems/trapping-rain-water/

class Solution {
public:
    int trap(vector<int>& height) {
        if(height.size() == 0)
            return 0;
        vector<int>left_arr(height.size());
        vector<int>right_arr(height.size());
        left_arr[0] = height[0];
        for(int i=1;i<height.size();i++)
        {
            left_arr[i] = max(left_arr[i-1], height[i]);
        }
        right_arr[height.size()-1] = height[height.size()-1];
        for(int j=height.size()-2; j>=0; j--)
        {
            right_arr[j] = max(right_arr[j+1], height[j]);
        }
        int ans = 0;
        for(int k=0; k<height.size(); k++)
            ans += min(right_arr[k], left_arr[k]) - height[k];
        return ans; 
    }
};
